### PR TITLE
#changed - Remove GeneratedColumn from 'Copy as SQL Insert' #1282

### PR DESCRIPTION
**Pull request by gduh**
## Changes:
- Remove GeneratedColumn from 'Copy as SQL Insert' 

## Tested:
- Processors:
  - [X] Intel
  - [ ] Apple Silicon
- macOS Versions:
  - [ ] 10.12.x (Sierra)
  - [ ] 10.13.x (High Sierra)
  - [ ] 10.14.x (Mojave)
  - [ ] 10.15.x (Catalina)
  - [X] 11.x (Big Sur)
  - [ ] 12.x (Monterey)
- Localizations:
  - [X] English
  - [ ] Spanish
  - [ ] Other (please specify)
- Xcode Version:13.0 (13A233)

### CODE REVIEW

File: Source/Views/TableViews/SPCopyTable.m [hunk 1]
```diff
@@ -403,11 +403,11 @@ - (NSString *)rowsAsCsvStringWithHeaders:(BOOL)withHeaders onlySelectedRows:(BOO
  */
 - (NSString *)rowsAsSqlInsertsOnlySelectedRows:(BOOL)onlySelected{
 	
-	return [self rowsAsSqlInsertsOnlySelectedRows:onlySelected skipAutoIncrementColumn:NO];
+	return [self rowsAsSqlInsertsOnlySelectedRows:onlySelected skipAutoIncrementColumn:NO skipGeneratedColumn:NO];
```
Comment (Jason-Morcos) [inline Source/Views/TableViews/SPCopyTable.m:406]: Why not skipGeneratedColumn:YES here?

File: Source/Views/TableViews/SPCopyTable.m [hunk 2]
```diff
@@ -448,49 +448,58 @@ - (NSString *)rowsAsSqlInsertsOnlySelectedRows:(BOOL)onlySelected skipAutoIncrem
...
-                if(autoIncrement == NO){
-                    autoIncrement = [obj boolForKey:@"AUTO_INCREMENT_FLAG"];
+        NSString *g = [[columnDefinitions safeObjectAtIndex:columnMappings[c]] objectForKey:@"generatedalways"];
+
+        // Generated column
+        if (skipGeneratedColumn == YES && g) {
+            [tbHeader removeObject:[[columnDefinitions safeObjectAtIndex:columnMappings[c]] objectForKey:@"name"]];
+            columnTypes[c] = 5;
+        } else {
+        // AutoInc column
```
Comment (Jason-Morcos) [inline Source/Views/TableViews/SPCopyTable.m:458]: I think removing the column from the tbHeader here can introduce errors below with the tbHeader:safeObjectAtIndex call for later columns.
Why do you need to call removeObject here at all? I believe setting the column type alone should be sufficient to create the desired behavior?

File: Source/Views/TableViews/SPCopyTable.m [hunk 3]
```diff
@@ -603,6 +612,10 @@ - (NSString *)rowsAsSqlInsertsOnlySelectedRows:(BOOL)onlySelected skipAutoIncrem
 					case 4:
 						SPLog(@"Not adding auto inc value");
 						break;
+                    // Generated column
```
Comment (Jason-Morcos) [inline Source/Views/TableViews/SPCopyTable.m:615]: Maybe tweak the comment to be something like "Generated column AND skip generated columns=YES" to be a little more clear that type 5 only exists if skip generated is YES

**Review by Jason-Morcos — CHANGES_REQUESTED** [review]
Good overall, have a couple questions/concerns but I think it's really close!

File: Source/Views/TableViews/SPCopyTable.m (hunk 1 above)
Comment (gduh) [reply Source/Views/TableViews/SPCopyTable.m:406]: It was just to leave the previous behavior for other functions than "Copy as SQL insert" to avoir any issue.

File: Source/Views/TableViews/SPCopyTable.m (hunk 2 above)
Comment (gduh) [reply Source/Views/TableViews/SPCopyTable.m:458]: Removing the column from the tbHeader is required here (Line 541-543): 

```
    // Begin the SQL string
    [result appendFormat:@"INSERT INTO %@ (%@)\nVALUES\n",
        [(selectedTable == nil) ? @"<table>" : selectedTable backtickQuotedString], [tbHeader componentsJoinedAndBacktickQuoted]];
```

It should not generate an error, since we are using a pointer, not an index to reference the object in the array. More over the safeObjectAtIndex method is safer than ObjectAtIndex.

File: Source/Views/TableViews/SPCopyTable.m (hunk 3 above)
Comment (gduh) [reply Source/Views/TableViews/SPCopyTable.m]: It's done

File: Source/Views/TableViews/SPCopyTable.m (hunk 2 above)
Comment (Jason-Morcos) [reply Source/Views/TableViews/SPCopyTable.m:458]: When you call removeObject on an NSMutableArray (tbHeader), that should cause indexes after the generated column to jump up by 1, correct? So, if the generated column(s) aren't at the end of the table, that later call (`[tbHeader safeObjectAtIndex:columnMappings[c]]` around 576) will result in the wrong table header being returned for columns in the table that come after the generated columns, correct?
In other words, the safe object for index thing isn't the issue, the issue is the shift in array indexes.
If you can test and verify that the correct insert statement is actually generated for tables that have non-generated columns after generated columns, then I'll feel better about this PR but I don't see how it would work correctly in that case

Comment (Jason-Morcos) [reply Source/Views/TableViews/SPCopyTable.m:458]: We could convert tbHeader into a dictionary perhaps and then not have issues with re-keying?
Perhaps I'm not reading the code correctly also, but it really seems like it would break down in the case described above.

Overall, the whole function is sorta hacky. Perhaps it could be refactored further to be simpler down the road.

Comment (gduh) [reply Source/Views/TableViews/SPCopyTable.m:458]: You're right, `[tbHeader safeObjectAtIndex:columnMappings[c]]` around 576 could be an issue. It occurs also when auto increment is removed (`[tbHeader removeObject:autoIncrementColumnName];`around 524.
But the error only gets thrown when we try to reload data and in my testing that part of code was never used. But, historically, I think if it's here it must be useful ?!

I'm agree the whole function is not very clean. I'll try to refactored it !

Comment (Jason-Morcos) [reply Source/Views/TableViews/SPCopyTable.m:458]: I don't have a general idea of how to make it better, except maybe it's worth using dictionaries/objects instead of having multiple arrays and using indexes to keep them all in sync :)

Comment (Jason-Morcos) [reply Source/Views/TableViews/SPCopyTable.m:458]: @gduh Any progress on this fix? I'd love to try to get it out in a release soon!

Comment (gduh) [reply Source/Views/TableViews/SPCopyTable.m:458]: > @gduh Any progress on this fix? I'd love to try to get it out in a release soon!

@Jason-Morcos  I'm trying to push this fix before the end of this week, ok for you ?

Comment (Jason-Morcos) [reply Source/Views/TableViews/SPCopyTable.m:458]: Works for me! Hoping to push a release with a bunch of fixes very beginning of 2022. Thank you for your work!

**Comment by gduh** [comment]
Refactored function : 
- Method is shorter
- Instead of added all columns first and then removing unwanted, add only the desired columns.
- use less loops
- Refactoring part of the code to try to be more readable
- Use an array of dictionary instead of severals array of different types, also fixes the historical bug about bad index.

**Review by Jason-Morcos — APPROVED** [review]
This looks really good! I'd like to get one more set of eyes on it because of how complex the logic is in general but it looks really good to me!

**Review by Kaspik — APPROVED** [review]
